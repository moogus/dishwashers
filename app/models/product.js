export const buildModel = (product = {}) => {
  const {
    title,
    media = {},
    price = {},
    details = {},
    displaySpecialOffer,
    additionalServices,
    code,
  } = product;

  const { now = '' } = price;
  const { productInformation } = details;
  const { images = {} } = media;
  const { altText = '', urls = [] } = images;

  return {
    title,
    media: { altText, urls },
    price: `Â£${now}`,
    productInformation,
    displaySpecialOffer,
    additionalServices,
    code,
  };
};
