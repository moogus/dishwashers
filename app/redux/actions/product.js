export const PRODUCT_LOADING = 'PRODUCT_LOADING';
export const PRODUCT_LOADED = 'PRODUCT_LOADED';
export const PRODUCT_ERROR = 'PRODUCT_ERROR';

export const loadProduct = (productId) => async (dispatch, _, config) => {
  dispatch({ type: PRODUCT_LOADING });

  try {
    const response = await fetch(`${config.productApi}${productId}`);
    const product = await response.json();

    dispatch({ type: PRODUCT_LOADED, product });
  } catch (err) {
    dispatch({ type: PRODUCT_ERROR });
  }
};
