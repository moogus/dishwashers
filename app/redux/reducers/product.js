import {
  PRODUCT_LOADING,
  PRODUCT_LOADED,
  PRODUCT_ERROR,
} from 'actions/product';

const intialState = { product: {}, loading: true };

export default (state = intialState, action) => {
  switch (action.type) {
    case PRODUCT_LOADING: {
      return {
        ...state,
        loading: true,
      };
    }

    case PRODUCT_LOADED: {
      const { product = {} } = action;

      return {
        ...state,
        product: {
          ...product,
        },
        loading: false,
      };
    }

    case PRODUCT_ERROR: {
      return {
        ...state,
        loading: false,
        error: true,
      };
    }

    default:
      return state;
  }
};
