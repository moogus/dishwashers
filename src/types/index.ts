export type GridItem = {
  productId: string;
  price: string;
  title: string;
  image: string;
};

export type AdditionalServices = {
  includedServices: string[];
}

export type Product = {
  title: string;
  media: Images,
  price: string;
  productInformation: string;
  displaySpecialOffer: string;
  additionalServices: AdditionalServices;
  code: string;
}

export type Price = {
  now?: string
}

export type ApiGridItem = {
  productId: string;
  price: Price;
  title: string;
  image: string;
}

export type Images = {
  altText: string;
  urls: string[]
}

export type Media = {
  images: Images;
}

export type Details = {
  productInformation: string;
}

export type ApiProduct = {
  title : string;
  media: Media;
  price?: Price;
  details?: Details;
  displaySpecialOffer : string;
  additionalServices: AdditionalServices;
  code : string;
}