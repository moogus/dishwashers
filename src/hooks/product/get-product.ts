import { useSelector } from "react-redux";
import { buildModel } from "../../models/product";

import {
  ApplicationState, 
  ProductsState
} from "../../interfaces";

export default () => {
  const { product } = useSelector<ApplicationState, ProductsState>((state) => state.product);
  return buildModel(product);
};
