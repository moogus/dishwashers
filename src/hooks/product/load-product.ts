import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { loadProduct } from "../../redux/actions/product";

import {
  ApplicationState, 
  ProductsState
} from "../../interfaces";

export default (productId) => {
  const dispatch = useDispatch();
  const { loading, error } = useSelector<ApplicationState, ProductsState>((state) => state.product);

  useEffect(() => {
    dispatch(loadProduct(productId));
    // eslint-disable-next-line
  }, []);

  return { loading, error };
};
