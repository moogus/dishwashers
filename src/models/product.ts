import { ApiProduct, Product } from "../types"

export const buildModel = (apiProduct:ApiProduct):Product => {
  const {
    title,
    media,
    price,
    details,
    displaySpecialOffer,
    additionalServices,
    code,
  } = apiProduct;

  const { now } = price;
  const { productInformation } = details;
  const { images } = media;
  const { altText, urls } = images;

  return {
    title,
    media: { altText, urls },
    price: `Â£${now}`,
    productInformation,
    displaySpecialOffer,
    additionalServices,
    code,
  };
};
