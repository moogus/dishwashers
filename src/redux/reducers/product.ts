import {
  PRODUCT_LOADING,
  PRODUCT_LOADED,
  PRODUCT_ERROR,
} from "../actions/product";

import { ApiProduct } from "../../types";
import { ProductsState } from "../../interfaces" 


export const intialState:ProductsState = { 
  product: {} as ApiProduct, 
  loading: true, 
  error: false
};

export default (state = intialState, action: { type?: string; product?: ApiProduct; }):ProductsState  => {
  switch (action.type) {
    case PRODUCT_LOADING: {
      return {
        ...state,
        loading: true,
      };
    }

    case PRODUCT_LOADED: {
      const { product } = action;

      return {
        ...state,
        product,
        loading: false,
      };
    }

    case PRODUCT_ERROR: {
      return {
        ...state,
        loading: false,
        error: true,
      };
    }

    default:
      return state;
  }
};
