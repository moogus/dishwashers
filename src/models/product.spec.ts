import { ApiProductModel } from "../types"

import { buildModel } from "./product";

describe("model is built from data", () => {
  test("hydrated model is build as expected", () => {
    expect(buildModel({
        title: "come a long way",
        media: {
          images: {
            alText: "foo",
            urls: ["1"],
          },
        },
        price: { now: "right here" },
        details: {},
        displaySpecialOffer: "",
        additionalServices: {},
        code: "1234",
      } as unknown as ApiProductModel)).toMatchSnapshot();
  });
});
