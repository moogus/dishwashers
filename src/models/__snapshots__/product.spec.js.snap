// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model is built from data hydrated model is build as expected 1`] = `
Object {
  "additionalServices": Object {},
  "code": "1234",
  "displaySpecialOffer": "",
  "media": Object {
    "altText": undefined,
    "urls": Array [
      "1",
    ],
  },
  "price": "Â£right here",
  "productInformation": undefined,
  "title": "come a long way",
}
`;
